src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1013,31 +1021,85 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
+        {
+            _cachedFrozenSnapshot ??= ComputeFrozenSnapshot(cancellationToken);
+            return _cachedFrozenSnapshot;
+        }
+    }
 
     private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancellationToken)
     {
         var newIdToProjectStateMapBuilder = this.SolutionState.ProjectStates.ToBuilder();
         var newIdToTrackerMapBuilder = _projectIdToTrackerMap.ToBuilder();
 
+        using var _1 = ArrayBuilder<DocumentState>.GetInstance(out var documentsToRemove);
+        using var _2 = ArrayBuilder<DocumentState>.GetInstance(out var documentsToAdd);
+
         foreach (var projectId in this.SolutionState.ProjectIds)
         {
+            cancellationToken.ThrowIfCancellationRequested();
+
             // if we don't have one or it is stale, create a new partial solution
-            var tracker = GetCompilationTracker(projectId);
-            var newTracker = tracker.FreezePartialState(cancellationToken);
+            var oldTracker = GetCompilationTracker(projectId);
+            var newTracker = oldTracker.FreezePartialState(cancellationToken);
+            if (oldTracker == newTracker)
+                continue;
 
             Contract.ThrowIfFalse(newIdToProjectStateMapBuilder.ContainsKey(projectId));
-            newIdToProjectStateMapBuilder[projectId] = newTracker.ProjectState;
+
+            var oldProjectState = this.SolutionState.GetRequiredProjectState(projectId);
+            var newProjectState = newTracker.ProjectState;
+
+            newIdToProjectStateMapBuilder[projectId] = newProjectState;
             newIdToTrackerMapBuilder[projectId] = newTracker;
+
+            // Freezing projects can cause them to have an entirely different set of documents (since it effectively
+            // rewinds the project back to the last time it produced a compilation).  Ensure we keep track of the docs
+            // added or removed from the project states to keep the final filepath-to-documentid map accurate.
+            //
+            // Note: we only have to do this if the actual project-state changed.  If we were able to use the same
+            // instance (common if we already got the compilation for a project), then nothing changes with the set
+            // of documents.
+            //
+            // Examples of where the documents may absolutely change though are when we haven't even gotten a
+            // compilation yet.  In that case, the project transitions to an empty state, which means we should remove
+            // all its documents from the filePathToDocumentIdsMap.  Similarly, if we were at some in-progress-state we
+            // might reset the project back to a prior state from when the last compilation was requested, losing
+            // information about documents recently added or removed.
+
+            if (oldProjectState != newProjectState &&
+                oldProjectState.DocumentStates.Ids != newProjectState.DocumentStates.Ids)
```

Comment:
This also needs to look for additional documents and .editorconfig to be correct too.

Reply:
done.